### PR TITLE
Log traceback

### PR DESCRIPTION
**Pull request by ramanaditya**
## Description
Based upon the workaround suggested by Pieter Marsman
Created a `@log_traceback` decorator, to be internally used to print the stack trace, log the errors and raise the errors.

## Motivation and Context

**issue:** #1210 

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [ ] Test, CI, or build

## Component(s) if applicable
- [ ] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [x] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:

- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/server/api_server.py [hunk 1]
```diff
@@ -60,6 +61,28 @@
...
+            return func(*args, **kwargs)
+        except Exception:
+            exception_type, exception_instance, exception_traceback = sys.exc_info()
+            formatted_traceback = "".join(traceback.format_tb(exception_traceback))
+            exception_message = (
+                f"{formatted_traceback}\n{exception_type.__name__}: "
+                f"{exception_instance}"
+            )
+            logger.error(exception_message)
+            raise exception_type(exception_message)
```
Comment (flosincapite) [inline bentoml/server/api_server.py:81]: If you just want to re-raise the original exception, you can do

`except Exception as exc`

at line 73, then, here, just do

`raise exc`

File: bentoml/server/api_server.py [hunk 2]
```diff
@@ -60,6 +61,28 @@
...
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception:
+            exception_type, exception_instance, exception_traceback = sys.exc_info()
+            formatted_traceback = "".join(traceback.format_tb(exception_traceback))
+            exception_message = (
+                f"{formatted_traceback}\n{exception_type.__name__}: "
```
Comment (flosincapite) [inline bentoml/server/api_server.py:77]: Please add a test which exercises this code path.

**Comment by bojiang** [comment]
IMO the main purpose of the issue is to log errors that occurred in the API function user code. It helps users to debug.
For the statics endpoint and swagger endpoint, all error happened should be the BentoML's internal error.

**Comment by bojiang** [comment]
I just drafted a PR to demonstrate the solution in my mind.

**Comment by ramanaditya** [comment]
> I just drafted a PR to demonstrate the solution in my mind.
> Do you think it makes sense?

The solution we created are pretty similar, the difference is I created the decorator, I thought it would make the code cleaner, 
about the implementation, we can directly wrap any function with this decorator.

If it was meant to be used in one place, then the draft PR you created looks nice. But if we want to use it at multiple places then I feel like we should go ahead with the function.


> For the statics and swagger endpoint, all error happened should be the BentoML's internal error, and raised directly.

I was not sure which all function to wrap, so I was just making the test over them. Please let me know the functions which we need to print traceback so that I could wrap them.

File: bentoml/server/api_server.py [hunk 3]
```diff
@@ -141,6 +165,7 @@ def static_serve(static_path, file_path):
             )
 
     @staticmethod
+    @log_traceback
```
Comment (parano) [inline bentoml/server/api_server.py:168]: this is probably not necessary, it is only meant for wrapping around user-defined API callback function.

Comment (parano) [reply bentoml/server/api_server.py:168]: no need to wrap it around static_serve nor index_view_func, I think @bojiang's suggestion makes sense. Using a decorator does not seem to make the code cleaner in this case

**Comment by yubozhao** [comment]
Close this with #1340 merged.

**Pull request closed without merging**